File: c/2015/15/main.c
#include "../../util.h"

#if 0
static char *FILE_NAME = "sample.txt";
#else
static char *FILE_NAME = "input.txt";
#endif

int64_t s(int8_t arr[][5], int64_t l, int8_t idx, int8_t coef[]) {
    int64_t ret = 0;
    for (int64_t i = 0; i < l; i++) {
        ret += arr[i][idx]*coef[i];
    }
    return ret > 0 ? ret : 0;
}

void part1() {
    ARR_MUL_STRING* in = readInputDelim(FILE_NAME, " ,");
    int8_t arr[4][5];
    for (int8_t i = 0; i < in->length; i++) {
        for (int8_t j = 0; j < 5; j++) {
            arr[i][j] = atoi(in->strings[i].strings[2+2*j]);
        }
    }

    uint64_t max = 0;
    for (int8_t a = 0; a < 100; a++) {
        for (int8_t b = 0; b < 100; b++) {
            for (int8_t c = 0; c < 100; c++) {
                for (int8_t d = 0; d < 100; d++) {
                    if (a+b+c+d != 100) continue;
                    int64_t val[4];
                    uint64_t tmp = 1;
                    int8_t coef[4] = {a,b,c,d};
                    for (int8_t i = 0; i < 4; i++) {
                        val[i] = s(arr, in->length, i, coef);
                        tmp *= val[i];
                    }
                    if (tmp > max) {
                        max = tmp;
                    }
                }
            }
        }
    }
    printf("Part 1: %ld\n", max);
    freeArrMulString(in);
}

void part2() {
    ARR_MUL_STRING* in = readInputDelim(FILE_NAME, " ,");
    int8_t arr[4][5];
    for (int8_t i = 0; i < in->length; i++) {
        for (int8_t j = 0; j < 5; j++) {
            arr[i][j] = atoi(in->strings[i].strings[2+2*j]);
        }
    }

    uint64_t max = 0;
    for (int8_t a = 0; a < 100; a++) {
        for (int8_t b = 0; b < 100; b++) {
            for (int8_t c = 0; c < 100; c++) {
                for (int8_t d = 0; d < 100; d++) {
                    if (a+b+c+d != 100) continue;
                    int8_t coef[4] = {a,b,c,d};
                    int64_t cal = s(arr, in->length, 4, coef);
                    if (cal != 500) continue;
                    int64_t val[4];
                    uint64_t tmp = 1;
                    for (int8_t i = 0; i < 4; i++) {
                        val[i] = s(arr, in->length, i, coef);
                        tmp *= val[i];
                    }
                    if (tmp > max) {
                        max = tmp;
                    }
                }
            }
        }
    }
    printf("Part 2: %ld\n", max);
    freeArrMulString(in);
}

int main(void) {
    part1();
    part2();
    return 0;
}

File: python/2015/15/main.py
#!/usr/bin/env python3

import sys
sys.path.append("../../")
sys.path.append("../")
import util as u

###############
debug = False
###############

file = "sample.txt" if debug else "input.txt"

def read_input():
    return [(int(x[0]), int(x[1]), int(x[2]), int(x[3]), int(x[4])) for x in u.read_match(file,".* (-?[0-9]+).* (-?[0-9]+).* (-?[0-9]+).* (-?[0-9]+).* (-?[0-9]+)")]

def mul(arr):
    ret = 1
    for a in arr:
        ret *= a
    return ret

def s(A, idx, n):
    ret = 0
    for i,x in enumerate(A):
        t = x[idx]*n[i]
        ret += t
    return ret if ret > 0 else 0

def part1():
    A = read_input()
    max = 0
    for a in range(100):
        for b in range(100):
            for c in range(100):
                for d in range(100):
                    if a+b+c+d != 100:
                        continue
                    g = [s(A, i, [a, b, c, d]) for i in range(4)]
                    x = mul(g)
                    if x > max:
                        max = x
    print("Part 1:",max)

def part2():
    A = read_input()
    max = 0
    for a in range(100):
        for b in range(100):
            for c in range(100):
                for d in range(100):
                    if a+b+c+d != 100:
                        continue
                    cal = s(A, 4, [a,b,c,d])
                    if cal != 500:
                        continue
                    g = [s(A, i, [a, b, c, d]) for i in range(4)]
                    x = mul(g)
                    if x > max:
                        max = x
    print("Part 2:",max)

    

part1()
part2()
